### PR TITLE
MULTIARCH-3676: PowerVS add transit gateway destroy

### PR DESCRIPTION
**Pull request by hamzy**
This is part of the new work for supporting Power Edge Router (PER).
This updates the destroy logic to also cleanup matching PER components in the cloud.

https://issues.redhat.com/browse/MULTIARCH-3676

### CODE REVIEW

File: pkg/destroy/powervs/cloud-transit-gateways.go [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+				listTransitGatewaysOptions.SetStart(*gatewayCollection.Next.Start)
+			} else {
+				o.Logger.Debugf("listTransitGateways: Next = nil")
+			}
+			moreData = gatewayCollection.Next != nil
+			o.Logger.Debugf("listTransitGateways: moreData = %v", moreData)
+		}
+	}
+
+	return cloudResources{}.insert(result...), nil
```
Comment (AshwinHIBM) [inline pkg/destroy/powervs/cloud-transit-gateways.go:118]: A bit confused why we're listing again in the Not Found case when we did so around L66 - L82.

Comment (hamzy) [reply pkg/destroy/powervs/cloud-transit-gateways.go:118]: Like every other part of the destroy code, we only print out the matching TGs here:
https://github.com/hamzy/installer/blob/PowerVS-Add-transit-gateway-destroy/pkg/destroy/powervs/cloud-transit-gateways.go#L55
If no matching TGs were found, then we print out every TG here:
https://github.com/hamzy/installer/blob/PowerVS-Add-transit-gateway-destroy/pkg/destroy/powervs/cloud-transit-gateways.go#L97

File: pkg/destroy/powervs/cloud-transit-gateways.go [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+		o.Logger.Debugf("listTransitGateways: NO matching transit gateway against: %s", o.InfraID)
+
+		listTransitGatewaysOptions = o.tgClient.NewListTransitGatewaysOptions()
+		listTransitGatewaysOptions.Limit = &perPage
+		moreData = true
+
+		for moreData {
+			gatewayCollection, response, err = o.tgClient.ListTransitGatewaysWithContext(ctx, listTransitGatewaysOptions)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to list transit gateways: %v and the respose is: %s", err, response)
```
Comment (r4f4) [inline pkg/destroy/powervs/cloud-transit-gateways.go:94]:
```suggestion
				return nil, fmt.Errorf("failed to list transit gateways: %w and the respose is: %s", err, response)
```

File: pkg/destroy/powervs/cloud-transit-gateways.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+			o.errorTracker.suppressWarning(item.key, err2, o.Logger)
+			return false, err2
+		})
+		if err != nil {
+			o.Logger.Fatalf("destroyTransitGateway: ExponentialBackoffWithContext (destroy) returns %v", err)
+		}
+	}
+
+	if items = o.getPendingItems(transitGatewayConnectionTypeName); len(items) > 0 {
+		return errors.Errorf("destroyTransitGateway: %d undeleted items pending", len(items))
```
Comment (r4f4) [inline pkg/destroy/powervs/cloud-transit-gateways.go:174]:
```suggestion
		return fmt.Errorf("destroyTransitGateway: %d undeleted items pending", len(items))
```

File: pkg/destroy/powervs/cloud-transit-gateways.go [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+			o.errorTracker.suppressWarning(item.key, err2, o.Logger)
+			return false, err2
+		})
+		if err != nil {
+			o.Logger.Fatalf("destroyTransitGateways: ExponentialBackoffWithContext (destroy) returns %v", err)
+		}
+	}
+
+	if items = o.getPendingItems(transitGatewayTypeName); len(items) > 0 {
+		return errors.Errorf("destroyTransitGateways: %d undeleted items pending", len(items))
```
Comment (r4f4) [inline pkg/destroy/powervs/cloud-transit-gateways.go:404]:
```suggestion
		return fmt.Errorf("destroyTransitGateways: %d undeleted items pending", len(items))
```

**Review by r4f4 — CHANGES_REQUESTED** [review]
Please squash the v4 -> v5 changes into one commit and separate the vendor changes into their own commit.

File: pkg/destroy/powervs/cloud-transit-gateways.go [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+	listTransitGatewaysOptions = o.tgClient.NewListTransitGatewaysOptions()
+	listTransitGatewaysOptions.Limit = &perPage
+
+	result := []cloudResource{}
+
+	for moreData {
+		// https://github.com/IBM/networking-go-sdk/blob/master/transitgatewayapisv1/transit_gateway_apis_v1.go#L184
+		gatewayCollection, response, err = o.tgClient.ListTransitGatewaysWithContext(ctx, listTransitGatewaysOptions)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list transit gateways: %v and the respose is: %s", err, response)
```
[the remark this answers is not in the document]
Comment (hamzy) [reply pkg/destroy/powervs/cloud-transit-gateways.go]: Done!

File: pkg/destroy/powervs/cloud-transit-gateways.go (hunk 2 above)
Comment (hamzy) [reply pkg/destroy/powervs/cloud-transit-gateways.go]: Done!

File: pkg/destroy/powervs/cloud-transit-gateways.go (hunk 3 above)
Comment (hamzy) [reply pkg/destroy/powervs/cloud-transit-gateways.go]: Done!

File: pkg/destroy/powervs/cloud-transit-gateways.go (hunk 4 above)
Comment (hamzy) [reply pkg/destroy/powervs/cloud-transit-gateways.go]: Done!

**Review by r4f4 — COMMENTED** [review]
/approve

**Pull request merged**